sources/libs/ud/parse/type.c

```diff
@@ -0,0 +1,89 @@
+#include <brutal/debug.h>
+#include <ud/parse/decl.h>
+#include <ud/parse/lexer.h>
+#include "brutal/parse/lex.h"
+#include "ud/ast/decl.h"
+#include "ud/ast/type.h"
+
+UdTupleDecl ud_parse_tuple_decl(Lex *lex, Alloc *alloc)
+{
+    UdTupleDecl ret = {};
+
+    vec_init(&ret.fields, alloc);
+
+    while (lex_curr(lex).type != UDLEX_RPAREN)
+    {
+        ud_parse_whitespace(lex);
+
+        UdType type = {.name = lex_curr(lex).str, .type = ud_str_to_type(lex_curr(lex).str)};
+
+        vec_push(&ret.fields, type);
+
+        lex_next(lex);
+
+        ud_parse_whitespace(lex);
+
+        if (lex_peek(lex, 1).type != UDLEX_RPAREN)
+        {
+            ud_parse_whitespace(lex);
+
+            if (lex_curr(lex).type != UDLEX_RPAREN)
+            {
+                ud_parse_whitespace(lex);
+
+                ud_expect(lex, UDLEX_COMMA);
+            }
+
+            else
+            {
+                break;
+            }
+        }
+
+        else
+        {
+            ud_parse_whitespace(lex);
+
+            lex_next(lex);
+
+            ud_parse_whitespace(lex);
+        }
+
+        if (ud_get_error())
+        {
+            break;
+        }
+    }
+
+    log$("{}", lex_curr(lex).str);
+
+    return ret;
+}
+
+UdDecl ud_parse_type_decl(Lex *lex, Alloc *alloc)
+{
+    UdDecl ret = {};
+
+    ret.type = UD_DECL_TYPE;
+
+    ud_parse_whitespace(lex);
+
+    if (ud_expect(lex, UDLEX_IDENT))
+    {
+        ret.name = lex_peek(lex, -1).str;
+    }
+
+    ud_parse_whitespace(lex);
+
+    ud_expect(lex, UDLEX_EQUAL);
+
```

Comment:
maybe do like cc, add a ud_expect_spaced() (I don't remember the name) that is just;
```c
void ud_expect_spaced(Lex* lex, ClexType type)
{
     ud_parse_whitespace(lex);
     ud_expect(lex, UDLEX_EQUAL);
     ud_parse_whitespace(lex);
}
```

Reply:
oh wait no you have ud_parse_expect_separator